src/PostgREST/App.hs

```diff
@@ -9,430 +9,549 @@ Some of its functionality includes:
+          ]
+
+    failNotSingular ctxContentType rQueryTotal $
+      if iPreferRepresentation == Types.Full then
+        response HTTP.status201 (headers ++ contentTypeHeaders context) (toS rBody)
+      else
+        response HTTP.status201 headers mempty
+
+handleUpdate :: QualifiedIdentifier -> RequestContext -> DbHandler Wai.Response
+handleUpdate identifier context@(RequestContext _ _ ApiRequest{..} contentType) = do
+  CreateResult{..} <- writeQuery identifier False mempty context
+
+  let
+    response = gucResponse rGucStatus rGucHeaders
+    fullRepr = iPreferRepresentation == Types.Full
+    updateIsNoOp = Set.null iColumns
+    status
+      | rQueryTotal == 0 && not updateIsNoOp = HTTP.status404
+      | fullRepr = HTTP.status200
+      | otherwise = HTTP.status204
+    contentRangeHeader =
+      RangeQuery.contentRangeH 0 (rQueryTotal - 1) $
+        if shouldCount iPreferCount then Just rQueryTotal else Nothing
+
+  failNotSingular contentType rQueryTotal $
+    if fullRepr then
+      response status (contentTypeHeaders context ++ [contentRangeHeader]) (toS rBody)
+    else
+      response status [contentRangeHeader] mempty
+
+handleSingleUpsert :: QualifiedIdentifier -> RequestContext-> DbHandler Wai.Response
+handleSingleUpsert identifier context@(RequestContext _ _ ApiRequest{..} _) = do
+  when (iTopLevelRange /= RangeQuery.allRange) $
+    throwError Error.PutRangeNotAllowedError
+
+  CreateResult{..} <- writeQuery identifier False mempty context
+
+  let response = gucResponse rGucStatus rGucHeaders
+
+  -- Makes sure the querystring pk matches the payload pk
+  -- e.g. PUT /items?id=eq.1 { "id" : 1, .. } is accepted,
+  -- PUT /items?id=eq.14 { "id" : 2, .. } is rejected.
+  -- If this condition is not satisfied then nothing is inserted,
+  -- check the WHERE for INSERT in QueryBuilder.hs to see how it's done
+  when (rQueryTotal /= 1) $ do
+    lift SQL.condemn
+    throwError Error.PutMatchingPkError
+
+  return $
+    if iPreferRepresentation == Types.Full then
+      response HTTP.status200 (contentTypeHeaders context) (toS rBody)
+    else
+      response HTTP.status204 (contentTypeHeaders context) mempty
+
+handleDelete :: QualifiedIdentifier -> RequestContext -> DbHandler Wai.Response
+handleDelete identifier context@(RequestContext _ _ ApiRequest{..} contentType) = do
+  CreateResult{..} <- writeQuery identifier False mempty context
+
+  let
+    response = gucResponse rGucStatus rGucHeaders
+    contentRangeHeader =
+      RangeQuery.contentRangeH 1 0 $
+        if shouldCount iPreferCount then Just rQueryTotal else Nothing
+
+  failNotSingular contentType rQueryTotal $
+    if iPreferRepresentation == Types.Full then
+      response HTTP.status200
+        (contentTypeHeaders context ++ [contentRangeHeader])
+        (toS rBody)
+    else
+      response HTTP.status204 [contentRangeHeader] mempty
+
+handleInfo :: Monad m => QualifiedIdentifier -> DbStructure -> Handler m Wai.Response
```

Comment:
I think the idiomatic solution would be to make our `Handler` a `ReaderT RequestContext` - then we wouldn't need to pass the context around everywhere, and the functions could `ask` for exactly what they need. It seems a bit more obscure then what we currently have, though, so I'd save that for a separate PR where we can discuss it.

I typed `handleInfo` as `Monad m => Handler m Wai.Response` to make it clear that `handleInfo` does not (and because of its type, cannot) access the database. Haven't found a better way to express it, happy for any suggestions. I would be reluctant to type it `DbHandler` just for consistency, as it's actually a special case.